smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/Parser.java

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.jmespath;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import software.amazon.smithy.jmespath.ast.AndExpression;
+import software.amazon.smithy.jmespath.ast.ComparatorType;
+import software.amazon.smithy.jmespath.ast.ComparisonExpression;
+import software.amazon.smithy.jmespath.ast.CurrentExpression;
+import software.amazon.smithy.jmespath.ast.ExpressionReferenceExpression;
+import software.amazon.smithy.jmespath.ast.FieldExpression;
+import software.amazon.smithy.jmespath.ast.FilterProjectionExpression;
+import software.amazon.smithy.jmespath.ast.FlattenExpression;
+import software.amazon.smithy.jmespath.ast.FunctionExpression;
+import software.amazon.smithy.jmespath.ast.IndexExpression;
+import software.amazon.smithy.jmespath.ast.LiteralExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectHashExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectListExpression;
+import software.amazon.smithy.jmespath.ast.NotExpression;
+import software.amazon.smithy.jmespath.ast.ObjectProjectionExpression;
+import software.amazon.smithy.jmespath.ast.OrExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.SliceExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+
+/**
+ * A top-down operator precedence parser (aka Pratt parser) for JMESPath.
+ */
+final class Parser {
+
+    /** The maximum binding power for a token that can stop a projection. */
+    private static final int PROJECTION_STOP = 10;
+
+    /** Tokens that can start an expression. */
+    private static final TokenType[] NUD_TOKENS = {
+            TokenType.CURRENT,
+            TokenType.IDENTIFIER,
+            TokenType.LITERAL,
+            TokenType.STAR,
+            TokenType.LBRACE,
+            TokenType.LBRACKET,
+            TokenType.FLATTEN,
+            TokenType.EXPREF,
+            TokenType.NOT,
+            TokenType.FILTER,
+            TokenType.LPAREN
+    };
+
+    /** Tokens that can follow led tokens. */
+    private static final TokenType[] LED_TOKENS = {
+            TokenType.DOT,
+            TokenType.LBRACKET,
+            TokenType.OR,
+            TokenType.AND,
+            TokenType.PIPE,
+            TokenType.LPAREN,
```

Comment:
Added a comment to explain this:

```
// While not found in the led() method, a led LPAREN is handled
// when parsing a nud identifier because it creates a function.
```